### ISSUE TITLE
Model crash of AWI-ESM-2.1 after PISM coupling

### ISSUE THREAD

**Issue by ackerlar** [esm-tools/esm_tools#445]
**Describe the bug**
After the coupling step from PISM to AWI-ESM-2.1 the model sometimes crashes with this error:
```
  0:  inquire_rerun_file: restart_125ka_awiesm-pism_spm.nc - No such file or directory
  0:  inquire_rerun_file: restart_125ka_awiesm-pism_glm.nc - No such file or directory
  0:  inquire_rerun_file: restart_125ka_awiesm-pism_g3bm.nc - No such file or directory
  0:    
  0:    
  0:  read_o3clim_4: fn0: ozon2014 fn1: ozon2015 fn2: ozon2016
  0:  number of pressure levels of ozone climatology: 66
  0:  pressure levels in [hPa]:
  0:  0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.01  
  0:  0.01    0.01    0.01    0.02    0.03    0.04    0.05    0.07    0.10    0.15  
  0:  0.20    0.30    0.40    0.50    0.70    1.00    1.50    2.00    3.00    4.00  
  0:  5.00    7.00   10.00   15.00   20.00   25.00   30.00   35.00   40.00   50.00 
  0:  60.00   70.00   80.00   90.00  100.00  115.00  130.00  150.00  170.00  200.00
  0:  250.00  285.00  300.00  350.00  400.00  450.00  500.00  600.00  650.00  700.00
  0:  750.00  780.00  800.00  850.00  925.00 1000.00
  0:                                                                                                                                                                                                               
  0: forrtl: severe (66): output statement overflows record, unit -5, file Internal Formatted Write 
```
`/work/ba0989/a270124/PalModII/experiments/125ka_awiesm-pism/log/125ka_awiesm-pism_awiesm_compute_20150101-20151231.log`

I don't think the crash is related to the missing restart files as the error ` inquire_rerun_file: restart_125ka_awiesm-pism_spm.nc - No such file or directory` also appears for example here:
`/work/ba0989/a270124/PalModII/experiments/125ka_awiesm-pism/log/125ka_awiesm-pism_awiesm_compute_20100101-20101231.log` but the simulation still went through.

And the restart files are only looked for in the years directly after the coupling step from PISM. So for example the no `inquire_rerun_file` error message here: `/work/ba0989/a270124/PalModII/experiments/125ka_awiesm-pism/log/125ka_awiesm-pism_awiesm_compute_20140101-20141231.log`


**System (please complete the following information):**
 - Supercomputer: mistral
 - Version
```
+---------------------+-------------+--------------------------------------------------------------------------------------------+--------------------------+----------------------------+
| package_name        | version     | file                                                                                       | branch                   | tags                       |
|---------------------+-------------+--------------------------------------------------------------------------------------------+--------------------------+----------------------------|
| esm_calendar        | 5.0.0       | /mnt/lustre01/pf/a/a270124/.local/lib/python3.6/site-packages                              |                          |                            |
| esm_database        | 5.0.0       | /mnt/lustre01/pf/a/a270124/.local/lib/python3.6/site-packages                              |                          |                            |
| esm_environment     | 5.1.1       | /mnt/lustre01/pf/a/a270124/.local/lib/python3.6/site-packages                              |                          |                            |
| esm_master          | 5.1.3       | /mnt/lustre01/pf/a/a270124/.local/lib/python3.6/site-packages                              |                          |                            |
| esm_motd            | 5.0.2       | /mnt/lustre01/pf/a/a270124/.local/lib/python3.6/site-packages                              |                          |                            |
| esm_parser          | 5.1.7       | /mnt/lustre01/pf/a/a270124/esm_parser                                                      | prep_release             | v5.1.7-17-g333f8ec         |
| esm_pism            | 0.0.1.dev15 | /mnt/lustre01/pf/a/a270124/.local/lib/python3.6/site-packages                              | main                     | Error                      |
| esm_plugin_manager  | 5.0.1       | /mnt/lustre01/pf/a/a270124/.local/lib/python3.6/site-packages                              |                          |                            |
| esm_profile         | 5.0.0       | /mnt/lustre01/pf/a/a270124/.local/lib/python3.6/site-packages                              |                          |                            |
| esm_rcfile          | 5.1.0       | /mnt/lustre01/pf/a/a270124/.local/lib/python3.6/site-packages                              |                          |                            |
| esm_runscripts      | 5.0.14      | /mnt/lustre01/pf/a/a270124/esm_runscripts                                                  | prep_release_awiesm2pism | v5.0.14-70-gd1dd278-dirty  |
| esm_tools           | 5.1.17      | /mnt/lustre02/work/ba0989/a270124/glacial-inception/software/github.com/ackerlar/esm_tools | prep_release_wiso        | v5.1.10-113-g2b4417e-dirty |
| esm_version_checker | 5.1.5       | /mnt/lustre01/pf/a/a270124/.local/lib/python3.6/site-packages                              |                          |                            |
+---------------------+-------------+--------------------------------------------------------------------------------------------+--------------------------+----------------------------+
```

**Comment by denizural** [comment]
Hi @ackerlar,

# TLDR: 


could you please change the line 
``` fortran
31   CHARACTER(len=256) :: message_text = ''         !++mgs
```
to something higher eg. 512 (or even more)
in file `mo_exception.f90` and build Echam again.

---



I looked at your Echam version (`/mnt/lustre01/pf/a/a270124/awiesm-2.1/echam-6.3.05p2-concurrent_radiation-paleodyn/src`). I am able to replicate the same error on a dummy program that I will show you later in the reply.

What I am speculating is that this is rather a Fortran problem. Here is my explanation. 

You added some debug messages to the file `echam/mo_clim.f90`:
```fortran
 49        yearnc%format = NETCDF
 50        write(*,*) "LA DEBUG: Open file ", ntslcl, ", yearnc = ", yearnc, ", IO_READ = ", IO_READ
 51        CALL IO_open_unit (ntslcl, yearnc, IO_READ)
 52        IO_file_id = yearnc%file_id
 53        write(*,*) "LA DEBUG: File opened"
```
According to the log file you posted above (`/work/ba0989/a270124/PalModII/experiments/125ka_awiesm-pism/log/125ka_awiesm-pism_awiesm_compute_20150101-20151231.log`), the code executes to line 50 but does not reach 53 since we don't see `LA DEBUG: File opened` in your log file.

Now, let's have a look at the subroutine `IO_open_unit` from `echam/mo_io.f90`. There are number of calls to the `message` subroutine such as 
```fortran
CALL message ('IO_open', 'file '//TRIM(fileinfo%file_name)//' already open')
```
`message` subroutine is defined in the file `mo_exception.f90` and note the limits
```fortran
CHARACTER(len=256) :: message_text = ''
...
CHARACTER(len=LEN(message_text)) :: write_text
```
On the other hand, we have the NetCDF filename length as `parameter (nf_max_name = 256)` (eg. from https://github.com/Unidata/netcdf-fortran/blob/f04bf72f4019ae768f981e4e30b3a522ee264b5a/fortran/netcdf3.inc)

What I am speculating is that if your filenames are very long (including the full path) then the `message` subroutine fails. Look at this dummy code:

```fortran
program main
    implicit none
    character(len=10) :: text
    write(text, '(a)') "0123456789"
    print *, text
    print *, "-> this works"

    write(text, '(a)') "01234567890"
    print *, text
    print *, "-> this won't get printed"
end program main
```
compile `ifort main.f90`, and run `./a.out` results in (`mistral`)
```
$ ./a.out
 0123456789
 -> this works
forrtl: severe (66): output statement overflows record, unit -5, file Internal Formatted Write
...
```
IMO, Echam message texts should have been declared to be longer since 256 characters is bit too short for most debug messages. It is funny that they declared the path_limit as `INTEGER, PARAMETER :: path_limit = 4096` but the messages that will contain these paths are much shorter.

**Comment by ackerlar** [comment]
Thank you @denizural Unfortunately, I am in the queue on mistral all day. I will test what you suggested and give an update once the run started.

**Comment by ackerlar** [comment]
The model crashed not with this error:
```
  0:  IO_open_unit: unit 92 allready assigned to basal_mass_flux_doubleing,basal_mass_flux_grounded,bheatflx,bmelt,climatic_mass_balance,dHdt,dbdt,flux_mag,ice_surface_temp,lat,lon,mask,shelfbmassflux,shelfbtemp,taub,taub_mag,tempbase,tempsurf,tendency_of_ice_amount,tendency_of_ice_amount_due_to_basal_ma
  0: 
  0: ================================================================================
  0: 
  0: FATAL ERROR in IO_open_unit: Run terminated.
  0:  FINISH called from PE: 0
```
These variables are actually fields in the PISM restart file ...
 
According to `IO_open` in `mo_io.f90` (which is called by `IO_open_unit`) this error message is called when `fileinfo%opened==.true.` To me it is not clear when and where `fileinfo` (which is `yearnc` in the call of `IO_open_unit` in `mo_clim.f90`) is assigned... So I guess `fileinfo%opened` should be false is it s when `yearnc` is initialized. It looks also quiet random to me as other runs are working fine.

The log file is here:
`/work/ba0989/a270124/PalModII/experiments/125ka_awiesm-pism/log/125ka_awiesm-pism_awiesm_compute_20150101-20151231.log.bak2`

**Comment by denizural** [comment]
Hi Lars,
I checked the Echam code in more detail and here are my explanations:

First of all, the line in the log file
`IO_open_unit: unit 92 allready assigned to basal_mass_flux_doubleing, .... # bunch of stuff`
is very suspicious. This line is printed here:

```fortran
! subroutine IO_open_unit in file mo_io.f90
229   SUBROUTINE IO_open_unit(unit, fileinfo, mode)
237     IF (fileinfo%opened) THEN
238       WRITE(message_text,'(a,i0,a,a)') 'IO_open_unit: unit ',unit, &
239            ' allready assigned to ', fileinfo%file_name
240       CALL message ('',message_text)
241       CALL finish  ('IO_open_unit', 'Run terminated.')
242     ENDIF
```
so `fileinfo%file_name` is not read correctly. I am not very aware of your AWI-ESM simulation setup but you can have a look at this. Maybe you can find the buggy point right away.

### fileinfo and yearnc variables:

`fileinfo` is a data structure and defined here:

```fortran
! MODULE mo_netcdf in file mo_netcdf.f90
128   TYPE FILE_INFO
129
130     LOGICAL :: opened                       ! open = .true. or closed = .false.
131     LOGICAL :: parallel                     ! parallel .true. else .false.
132     INTEGER :: file_id                      ! netCDF file id
133     INTEGER :: access_mode                  ! access mode for that file
134     INTEGER :: ncdims(NF_MAX_VAR_DIMS)
135     INTEGER :: FORMAT                       ! file format NETCDF
136
137     CHARACTER(len=NF_MAX_NAME) :: creation_program ! name of this program
138     CHARACTER(len=NF_MAX_NAME) :: creation_user    ! who has run this program
139     CHARACTER(len=NF_MAX_NAME) :: creation_date    ! created netCDF file at
140     CHARACTER(len=NF_MAX_NAME) :: binary_source    ! binary type (CRAY or IEEE)
141     CHARACTER(len=NF_MAX_NAME) :: file_type        ! initital or restart file
142     CHARACTER(len=NF_MAX_NAME) :: file_name        ! nc file name
143     CHARACTER(len=NF_MAX_NAME) :: title
144   END TYPE FILE_INFO
```

`yearnc` is a variable of type `FILE_INFO` as initialized here:
```fortran
! MODULE mo_io in file mo_io.f90
42   TYPE (FILE_INFO), SAVE         :: yearnc
```
`SAVE` attribute means that `yearnc` is a **static** variable. It is a persistent variable and when it goes out of the scope (eg. module or subroutine), the last set values are kept. The default variable type is automatic which means that when it enters the scope of a function, it gets re-initialized.

Here are the all occurences of `yearnc` in the Echam code:

```fortran
./mo_io.f90:42:  TYPE (FILE_INFO), SAVE         :: yearnc
./mo_io.f90:625:    yearnc%opened = .FALSE.

./mo_clim.f90:49:       yearnc%format = NETCDF
./mo_clim.f90:50:       write(*,*) "LA DEBUG: Open file ", ntslcl, ", yearnc = ", yearnc, ", IO_READ = ", IO_READ
./mo_clim.f90:51:       CALL IO_open_unit (ntslcl, yearnc, IO_READ)
./mo_clim.f90:52:       IO_file_id = yearnc%file_id
./mo_clim.f90:79:       CALL IO_close(yearnc)
./mo_clim.f90:125:       CALL IO_open_unit (nvltcl, yearnc, IO_READ)
./mo_clim.f90:126:       IO_file_id = yearnc%file_id
./mo_clim.f90:151:       CALL IO_close(yearnc)
./mo_clim.f90:193:       CALL IO_open_unit (nvgratcl, yearnc, IO_READ)
./mo_clim.f90:194:       IO_file_id = yearnc%file_id
./mo_clim.f90:219:       CALL IO_close(yearnc)
```

Where is it initialized?
Inside the `subroutine IO_info_construct` in the file `mo_netcdf.f90`: 
```fortran
204   SUBROUTINE IO_info_construct(fileinfo)
205     TYPE(FILE_INFO), INTENT(inout)  :: fileinfo
206
207     fileinfo%opened           = .FALSE.
208     fileinfo%parallel         = .FALSE.
209     fileinfo%file_id          = 0
210     fileinfo%access_mode      = 0
211     fileinfo%ncdims(:)        = 0
212     fileinfo%creation_program = '?'
213     fileinfo%creation_user    = '?'
214     fileinfo%creation_date    = '?'
215     fileinfo%binary_source    = '?'
216     fileinfo%file_type        = '?'
217     fileinfo%file_name        = '?'
218     fileinfo%title            = '?'
219
220   END SUBROUTINE IO_info_construct
```

This is the sketch of the call tree:
```
SUBROUTINE readtslclim
    write(*,*) "LA DEBUG: Open file "                ! line 50
    CALL IO_open_unit (ntslcl, yearnc, IO_READ)      ! line 51 
    SUBROUTINE IO_open_unit(unit, fileinfo, mode)
        call IO_open(filename, fileinfo, mode)
            call IO_info_construct(fileinfo)
```

What is strange is `yearnc%opened = .FALSE.` inside `SUBROUTINE IO_init` which is called as early as in `initialize.f90`. Why is it opened although it is initialized as `FALSE`?

A **TODO** for you is to place a debug line to `mo_io.f90` and print the file names `fileinfo%file_name`

```fortran
 204     IF (status == NF_NOERR) THEN
 205       fileinfo%opened = .TRUE.
```

It must be passed for unit 92 since this is the only place in Echam code where it turn the `opened` flag on. Let's try and see what it gives. But in any case there is a severe error with the file name.

Another useful tip for you is there are 2 useful macros `__LINE__` and `__FILE__`. During the compile time Fortran preprocessor will replace them with the line number and file name where they are defined.

I hope these will help you. Good luck 🤞

**Comment by ackerlar** [comment]
I think, I finally found the reason for this error: During the coupling (and especially when I had to resubmit the coupling job due to other errors) the latest ECHAM restart file is modified including adding the `history`and `history_of_appended_files` which get extreme long `/work/ba0989/a270124/PalModII/experiments/125ka_awiesm-test/restart/echam/restart_125ka_awiesm-test_10201231_echam.nc.bak` Here are then also the PISM commands that popped up during the debugging of this error. When I removed the `history` etc from the restart file I first got some other issues but now after starting from the previous year it semms to work.
I have adapted the coupling script and will close the issue.
